python-lib/osisoft_client.py

```diff
@@ -896,6 +918,88 @@ def close(self):
+    # When request buffer is full, it is flushed into the batch endpoint
+
+    # Each write_row call is adds an entry to the responses list
+    # The pointer into that list is past as a _dku_counter parameter, extracted from the requests just before sending the batch
+    # Uppon receiving the batch response, each individual reponse is reordered
+    #   and the status code / error messages are stored into the responses list at the right pointer
+
+    # Possible improvement: flush the responses list (and result writing in the recipe.py) at each _flush_requests to keep memory from going up.
+
+    # If you have to review this, please accept my sincere apologies.
+
+    def __init__(self, client, max_streak_buffer_size=500, max_requests_buffer_size=500):
+        logger.info("Initializing OSIsoftBatchWriter, msbs={}, mrbs={}".format(max_streak_buffer_size, max_requests_buffer_size))
+        self.client = client
+        self.streak_buffer = []     # list of points from a single webid -> can be all sent in one request
+        self.requests_buffer = []   # list of independant requests -> sent one batch at a time
+        self.responses = []         # building a list of reponses in same order as write_row calls
+        self.current_webid = None
+        self.max_buffer_size = max_streak_buffer_size
+        self.max_requests_buffer_size = max_requests_buffer_size
+        self.current_streak = 0
+        self.row_number = 0
+
+    def write_row(self, webid, timestamp, value):
+        response = None
+        timestamp = self._timestamp_convertion(timestamp)
+        # mark in self.responses that status of this write depends on result of streak X
+        if self.current_webid is None:
+            logger.info("webid now is {}".format(webid))
+            self.current_webid = webid
+        if webid != self.current_webid or len(self.streak_buffer) >= self.max_buffer_size:
+            logger.info("webid: {} / {}".format(webid, self.current_webid))
+            self._flush_streak()
+            self.current_webid = webid
+        self.streak_buffer.append(
+            {
+                "Timestamp": "{}".format(timestamp),
+                "Value": "{}".format(value),
+                "_dku_counter": self.row_number
+            }
+        )
+        self.row_number += 1
+        logger.info("Pushed in buffer, now {}".format(len(self.streak_buffer)))
+        self.responses.append({"streak": self.current_streak})
+        return response
+
+    def _flush_streak(self):
+        # streak buffer must be flushed every time the webid changes
+        logger.info("flushing streak")
+        kwargs = self.client.prepare_post_all_values(self.current_webid, self.streak_buffer)
+        self.requests_buffer.append(kwargs)
+        logger.info("pushed to requests buffer {}".format(len(self.requests_buffer)))
+        if len(self.requests_buffer) >= self.max_requests_buffer_size:
```

Comment:
Same here. Could be nice to have a clear log saying for e.g, `Buffer is full, flushing current requests`